### PR TITLE
[Evaluation] Custom evaluation notebook: a reliability classifier

### PR DESCRIPTION
**Pull request by kaisopos**
# Description

Custom evaluation notebook: a reliability classifier.

Missing features to be more efficient (by design; to be simpler):
- Evaluates with a local model which does NOT perform well at all.
- Conversation IDs (ANLI's `uid`) are not retained.
- Intermediate results are not captured.
- No bootstrapping / confidence intervals.
- Single evaluation metric (BACC).
- Definition of `is_unsupported`, `is_supported` has been simplified and does not capture corner cases. Original:
```
is_unsupported = "<|unsupported|>" in response or "<unsupported>" in response
is_supported = "<|supported|>" in response or "<supported>" in response
```

## Related issues

Towards OPE-1097

## Before submitting

- [ ] This PR only changes documentation. (You can ignore the following checks in that case)
- [x] Did you read the [contributor guideline](https://github.com/oumi-ai/oumi/blob/main/CONTRIBUTING.md) Pull Request guidelines?
- [x] Did you link the issue(s) related to this PR in the section above?
- [x] Did you add / update tests where needed?

## Reviewers

At least one review from a member of `oumi-ai/oumi-staff` is required.

### CODE REVIEW

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from pathlib import Path\n",
+    "\n",
+    "tutorial_dir = \"tutorial\"\n",
```
Comment (wizeng23) [inline notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:74]: Could you make this more descriptive? Ex. `custom_eval_tutorial` or `reliability_eval_tutorial`

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Dataset({\n",
+      "    features: ['uid', 'premise', 'hypothesis', 'label', 'reason'],\n",
+      "    num_rows: 1200\n",
+      "})\n"
+     ]
```
Comment (wizeng23) [inline notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:121]: Could you print out an example from the dataset here?

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+    "    prompt = prompt_template.format(\n",
+    "        premise=anli_example[\"premise\"], hypothesis=anli_example[\"hypothesis\"]\n",
+    "    )\n",
+    "    message = {\"role\": \"user\", \"content\": prompt}\n",
+    "    metadata = {\"label\": convert_label_to_binary(anli_example[\"label\"])}\n",
+    "    conversation = {\"messages\": [message], \"metadata\": metadata}\n",
+    "    evaluation_dataset.append(conversation)\n",
+    "\n",
+    "# Limit the number of examples, if requested.\n",
+    "if NUM_EXAMPLES:\n",
```
Comment (wizeng23) [inline notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:207]: nit: There's some wasted computation with this approach. Could you have the for loop run for only NUM_EXAMPLES iterations?

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+    "## Evaluation\n",
+    "\n",
+    "This section discusses how we can leverage Oumi's custom evaluation to run inference on the prompts, extract the predictions from the model responses, and finally calculate metrics to assess our classification quality."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Step 1: Set the Evaluation Configuration\n",
```
Comment (wizeng23) [inline notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:227]: nit: Could this be moved after steps 1 and 2? IMO that's better for ordering, as it's defined right before it's used in step 4, and you don't need to jump to step 3 like you do now. You could move the description text about the inference engine from step 2 to be in this section instead.

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+   "outputs": [],
+   "source": [
+    "from sklearn.metrics import balanced_accuracy_score\n",
+    "\n",
+    "from oumi.core.evaluation import EvaluationResult\n",
+    "from oumi.core.registry import register_evaluation_function\n",
+    "\n",
+    "\n",
+    "@register_evaluation_function(\"my_custom_evaluation\")\n",
+    "def my_custom_evaluation(task_params, config, inference_engine, dataset):\n",
```
Comment (wizeng23) [inline notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:314]: nit: add typing?

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+    }
+   ],
+   "source": [
+    "from oumi.core.configs import EvaluationConfig\n",
+    "from oumi.core.evaluation import Evaluator\n",
+    "\n",
+    "config = EvaluationConfig.from_yaml(str(Path(tutorial_dir) / \"custom_eval.yaml\"))\n",
+    "\n",
+    "evaluator = Evaluator()\n",
+    "result = evaluator.evaluate(config, dataset=MY_DATASET)"
```
Comment (wizeng23) [inline notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:379]: nit: consider adding typing to this for clarity,  i.e. `result: list[EvaluationResult] = ...`

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Step 5: Inspect the Results\n",
+    "\n",
+    "The code below shows how we can inspect the `result` (an `EvaluationResult` object) that is returned by the evaluation function and how to display the relevant metrics."
```
Comment (wizeng23) [inline notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:388]: nit: technically, result is a list of EvaluationResult

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb
Comment (wizeng23) [inline notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:1]: WDYT about mentioning custom evaluation in the notebook title? I feel like that's what more users would be interested in, vs. reliability classifiers which seems more specific. What about something like `notebooks/Oumi - Custom Evaluation.ipynb`? If you think reliability classification is actually a useful example for some people and not just a toy example, could consider finding a way to mention that in the title in addition to custom evaluation.

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 2 above)
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:121]: I want to avoid this; the main reason is because I want to keep the code minimal, so one can go though the notebook and feel the code is very little.

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 1 above)
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:74]: Will do

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 4 above)
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:227]: Yep, that makes sense - will do!

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 5 above)
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:314]: No, better not; avoiding it by design. 
I dont want to import too many stuff, I dont want the function to seem cluttered. Also, students never do this, they may get confused with the syntax.

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 6 above)
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:379]: Avoiding typing by design (see above)

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 3 above)
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:207]: +1 on the comment, but I am leaning towards keeping it as is for simplicity, which IMO is more important than code efficiency (for this particular notebook). 

I don't like this syntax much, which I assume is what you are suggesting (I prefer it the way it is):
```
for example_index in min(NUM_EXAMPLES , len(anli_dataset)):
     ... reference as: anli_dataset[example_index]["some variable"]
```

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:1]: I don't have a strong opinion here; this will be used to back-up a blog on reliability classifier comparisons but, still, the name doesn't have to include "reliability classifier". Will check with others too and update based on how the majority feels.

Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:1]: CC @jgreer013 , @taenin

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+    "\n",
+    "# Convert the ANLI dataset into a list of conversations (oumi.core.types.Conversation).\n",
+    "evaluation_dataset = []\n",
+    "for anli_example in anli_dataset:\n",
+    "    prompt = prompt_template.format(\n",
+    "        premise=anli_example[\"premise\"], hypothesis=anli_example[\"hypothesis\"]\n",
+    "    )\n",
+    "    message = {\"role\": \"user\", \"content\": prompt}\n",
+    "    metadata = {\"label\": convert_label_to_binary(anli_example[\"label\"])}\n",
+    "    conversation = {\"messages\": [message], \"metadata\": metadata}\n",
```
Comment (jgreer013) [inline notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:203]: Add the `uid` as the `conversation_id` to the conversation.

**Review by jgreer013 — APPROVED** [review]

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb
Comment (jgreer013) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:1]: Maybe change title to `Oumi - Custom Evaluation Example (Reliability Classifier)

Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:1]: Jeremy suggested `Oumi - Custom Evaluation (Reliability Classifier)`
@wizeng23 wdyt?

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 2 above)
Comment (wizeng23) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:121]: IMO we should prioritize it being helpful/educational over minimal. I think seeing an example of the dataset is pretty important to having an understanding of how the training works. You could even drop the `print(anli_dataset)` line if being minimal was that important. However, I think simply changing it to `print(anli_dataset[0])` achieves a lot more.

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 3 above)
Comment (wizeng23) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:207]: How much does it affect runtime?

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 5 above)
Comment (wizeng23) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:314]: I understand the sentiment, but I don't think it's a compelling enough reason personally. We're building a framework used by thousands of people, not a small research codebase that may only be used for one paper. I think we should be held to a higher standard than a student's codebase. Students often don't comment their code either :p (or at least I didn't as a student).

Typing is the convention throughout our codebase, and fairly common in Python, so IMO it's a thing students will become familiar with anyway. It dramatically helps readability too. The types of the arguments are obvious to us who've been working on this project for over half a year, but it won't be to a newcomer. I don't like the idea of looking at how the args are used to try to guess what type they are.

I could understand giving this some slack since it's a Colab and not core code. However, I'm personally in favor of Colabs being educational over minimal, and I think typing helps for that. We want people to contribute to our codebase, and to do so, they have to learn it first. Colabs are a natural way to do that.

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 7 above)
Comment (wizeng23) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:388]: I know I'm being annoying with my PR comments, but even if you're against type hints, we should not be giving wrong type hints :p

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb
Comment (wizeng23) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:1]: SGTM!

Comment (taenin) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:1]: How about something along the lines of `Oumi- Build your own Custom Evaluation (Reliability Classifier)`
or
`Oumi- Build your own Custom Benchmark (Reliability Classifier)`

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 1 above)
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb]: Done

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 4 above)
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb]: Done

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 7 above)
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb]: You are NOT annoying. Well-intended feedback is always a gift :)
Fixed this (and I am not against type hints; but, for this notebook's audience specifically, I am prioritizing simplicity over robustness)

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb (hunk 8 above)
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb]: Done

File: notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb
Comment (kaisopos) [reply notebooks/Oumi - Evaluation of a Reliability Classifier.ipynb:1]: Yep, that seems like a great option!
I ll go with "Oumi - Build your own Custom Evaluation (Reliability Classifier).ipynb"
